lib/puma/server.rb

```diff
@@ -595,6 +600,24 @@ def handle_request(req, lines)
       env[RACK_INPUT] = body
       env[RACK_URL_SCHEME] =  env[HTTPS_KEY] ? HTTPS : HTTP
 
+      if @early_hints
+        env[EARLY_HINTS] = lambda { |links|
+          fast_write client, "HTTP/1.1 103 Early Hints\r\n".freeze
+
+          links.each do |link|
+            fast_write client, "Link: <#{link[:link]}>; rel=preload"
+            if as = link[:as]
+              fast_write client, "; as=#{as}"
+            end
+            fast_write client, "\r\n".freeze
+          end
+
+          fast_write client, "\r\n".freeze
+        }
+      else
+        env[EARLY_HINTS] = lambda { |_| }
```

Comment:
Why is this necessary? I guess I don't understand.

Reply:
This is so that app code can run without changes even if the server is being run without early hints support.  The example app that @eileencodes put in the PR description can be run whether early hints is enabled or not (it prevents the app code from writing a `if env[EARLY_HINTS]` conditional)